server/user_db.go

```diff
@@ -0,0 +1,235 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/golang/protobuf/ptypes"
+
+	pb "github.com/googleapis/gapic-showcase/server/genproto"
+	"google.golang.org/genproto/protobuf/field_mask"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func NewUserDb() UserDb {
+	return &userDb{
+		uid:   &uniqID{},
+		token: NewTokenGenerator(),
+		mu:    sync.Mutex{},
+		keys:  map[string]dbIndex{},
+		users: []userEntry{},
+	}
+}
+
+type UserDb interface {
+	Create(u *pb.User) (*pb.User, error)
+	Get(name string) (*pb.User, error)
+	Update(u *pb.User, f *field_mask.FieldMask) (*pb.User, error)
+	Delete(name string) error
+	List(page_size int32, token string) (*pb.ListUsersResponse, error)
+}
+
+type userEntry struct {
+	user    *pb.User
+	deleted bool
```

Comment:
If we keep adding and removing, we'll eventually run out of memory. Probably not a big consideration here though?

Reply:
Agreed. Not too big of a consideration. I made the server rely on some DB interface for this reason. If ever we need to scale, we can just write a userDB interface to an actual database. I foresee any tests ever getting near causing an OOM error if the server is only alive for the current testing session.

Reply:
Furthermore, the OOM problem is inherent making an in-memory db. For this application the deleted bool is simply to provide a simple way of generating page tokens. Since nothing is ever removed from the list, we can rely on index for page tokens.